feat: add Pectra BLS12-381 elliptic curve precompiles

yelhousni opened the pull request:
# Description

<!-- Please include a summary of the changes and the related issue. Please also include relevant motivation and context. -->

This PR adds gnark circuits corresponding to precompiles for BLS12-381 curve operations https://eips.ethereum.org/EIPS/eip-2537.

- [x] BLS12_G1ADD
- [x] BLS12_G1MSM
- [x] BLS12_G2ADD
- [x] BLS12_G2MSM
- [x] BLS12_PAIRING_CHECK
- [x] BLS12_MAP_FP_TO_G1 (@ThomasPiellard)
- [x] BLS12_MAP_FP2_TO_G2 (https://github.com/Consensys/gnark/pull/1040)

TODO:
- [x] BLS12_PAIRING_CHECK: inlcude G2 memerbship in Miller loop (lines computation).
- [x] BLS12_G2ADD: check the two addends are on G2.
- [x] BLS12_G2MSM: optimized `ScalarMul()` with GLV 
    - [ ] GLS? fakeGLS? GLV/S+fakeGLV/S?
- [ ] BLS12_G1MSM and BLS12_G2MSM: optimize MSM algorithm.

## Type of change

<!-- Please delete options that are not relevant. -->

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

<!-- Please describe the tests that you ran or implemented to verify your changes. Provide instructions so we can reproduce. -->

Tests are implemented in `std/evmprecompiles/bls_test.go`.

# How has this been benchmarked?

<!-- Please describe the benchmarks that you ran to verify your changes. -->

PLONK (on BLS12-377) number of constraints (scs):

| precompile | curve membership test? | subgroup membership test? | constraints (scs) |
|------------|-------------------------|----------------------------|------------------|
|BLS12_G1ADD | Yes | No | 9 641 |
|BLS12_G2ADD | Yes | No | 23 668 |

| precompile | curve membership test? | subgroup membership test? | constraints (scs) |
|------------|-------------------------|----------------------------|------------------|
|BLS12_G1MSM  (1 pair) <br>*Scalar Multiplication*<br> | Yes | Yes | 590 418 |
|BLS12_G1MSM  (2 pairs) | Yes | Yes | 1 132 727 |
|BLS12_G1MSM  (3 pairs) | Yes | Yes | 1 569 303 |
|BLS12_G1MSM  (5 pairs) | Yes | Yes | 2 442 455 |
|BLS12_G1MSM  (10 pairs) | Yes | Yes | 4 625 335 |
|BLS12_G1MSM  (20 pairs) | Yes | Yes | 8 991 095 |
|BLS12_G1MSM  (100 pairs) | Yes | Yes | 43 917 175 |

| precompile | curve membership test? | subgroup membership test? | constraints (scs) |
|------------|-------------------------|----------------------------|------------------|
|BLS12_G2MSM  (1 pair) <br>*Scalar Multiplication*<br> | Yes | Yes | 1 225 743 |
|BLS12_G2MSM  (2 pairs) | Yes | Yes | 2 196 208 |
|BLS12_G2MSM  (3 pairs) | Yes | Yes | 3 166 673 |
|BLS12_G2MSM  (5 pairs) | Yes | Yes | 5 107 603 |
|BLS12_G2MSM  (10 pairs) | Yes | Yes | 9 959 981 |
|BLS12_G2MSM  (20 pairs) | Yes | Yes | 19 665 567 |
|BLS12_G2MSM  (50 pairs) | Yes | Yes | 48 794 236 |

| precompile | curve membership test? | subgroup membership test? | constraints (scs) |
|------------|-------------------------|----------------------------|------------------|
|BLS12_PAIRING_CHECK  (2 pairs) | Yes | Yes | 2 821 323 |
|BLS12_PAIRING_CHECK  (4 pairs) | Yes | Yes | 4 993 571 |
|BLS12_PAIRING_CHECK  (9 pairs) | Yes | Yes | 10 424 481 |
|BLS12_PAIRING_CHECK  (20 pairs) | Yes | Yes | 22 371 551 |

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

ivokub commented on std/algebra/emulated/sw_bls12381/g1.go:
```diff
@@ -157,6 +168,55 @@ func (g1 *G1) scalarMulBySeedSquare(q *G1Affine) *G1Affine {
+	b := g1.curveF.Select(selector, g1.curveF.Zero(), &four)
+
+	left = g1.curveF.Mul(&P.Y, &P.Y)
+	right = g1.curveF.Mul(&P.X, &P.X)
```
could use `g1.curve.Eval` to reduce multiplications

ivokub commented on std/evmprecompiles/12-blsg1msm.go:
```diff
@@ -0,0 +1,36 @@
+
+	// Check that Pᵢ are on G1
+	for _, p := range P {
+		g1.AssertIsOnG1(p)
```
We can also use generic `Pairing` interface from `algebra` and use its AssertIsOnG1. But in the current context we only care for BLS12-381, so can continue using the current approach.

ivokub commented on std/evmprecompiles/15-blspairing.go:
```diff
@@ -0,0 +1,114 @@
+		panic(err)
+	}
+	for i := 0; i < n; i++ {
+		// 1- Check that Pᵢ are on G1
```
I would add explicitly the comment why in this case we do it here, not in the zkevm. Then imo better to review later when comparing to bn254 pairing precompile

ivokub commented on std/evmprecompiles/bls_test.go:
```diff
@@ -0,0 +1,310 @@
+		}, ecc.BN254.ScalarField())
+		assert.NoError(err)
+	}
+}
```
Missing map to G1 test.

yelhousni replied:
@ThomasPiellard

ivokub reviewed (changes requested):
In principle looks good. But had a few comments about optimizing some computations and refactoring a bit.

Map to G1 is still ongoing. I'll rewrite a bit to have less duplication and better interfaces. Also need to add tests in the evmprecompiles packages.

yelhousni commented on std/algebra/emulated/sw_bls12381/g1.go:
```diff
@@ -157,6 +168,55 @@ func (g1 *G1) scalarMulBySeedSquare(q *G1Affine) *G1Affine {
+	g1.curveF.AssertIsEqual(left, right)
+}
+
+func (g1 *G1) AssertIsOnG1(P *G1Affine) {
```
it's in `TestGroupMembershipSolve`

yelhousni commented on std/algebra/emulated/sw_bls12381/g2.go:
```diff
@@ -119,6 +142,61 @@ func (g2 *G2) scalarMulBySeed(q *G2Affine) *G2Affine {
+// [BriJoy02]: https://link.springer.com/content/pdf/10.1007/3-540-45664-3_24.pdf
+// [EVM]: https://ethereum.github.io/yellowpaper/paper.pdf
+func (g2 *G2) AddUnified(p, q *G2Affine) *G2Affine {
+
```
I tried earlier with `Eval` but it didn't help because here we re-use `p.x+q.x`

yelhousni commented on std/algebra/emulated/sw_bls12381/g2.go:
```diff
@@ -276,8 +354,459 @@ func (g2 G2) doubleAndAdd(p, q *G2Affine) *G2Affine {
+// [ELM03]: https://arxiv.org/pdf/math/0208038.pdf
+// [EVM]: https://ethereum.github.io/yellowpaper/paper.pdf
+// [Joye07]: https://www.iacr.org/archive/ches2007/47270135/47270135.pdf
+func (g2 *G2) scalarMulGeneric(p *G2Affine, s *Scalar, opts ...algopts.AlgebraOption) *G2Affine {
```
yup it was used before I got the glv working. We can keep for testing glv or remove it.

ivokub replied:
I guess it doesn't hurt to have it as a compatibility test

ivokub commented:
I added https://github.com/Consensys/gnark-crypto/pull/674 - I'll go over G2 map to curve to see that it is also useful for https://github.com/Consensys/gnark/pull/1040.

Otherwise the map to curve also looks good. @yelhousni - please re-request review when you have addressed all the comments and I'll re-review. I had to merge master because there were some changes due to gnark-crypto API changes.

yelhousni commented:
> I added [Consensys/gnark-crypto#674](https://github.com/Consensys/gnark-crypto/pull/674) - I'll go over G2 map to curve to see that it is also useful for #1040.
> 
> Otherwise the map to curve also looks good. @yelhousni - please re-request review when you have addressed all the comments and I'll re-review. I had to merge master because there were some changes due to gnark-crypto API changes.

Addressed all the points except the `AssertIsOnCurve` and `AssertIsOnG1` in the `std/algebra.Curve` interface. No strong opinion here.

ivokub commented:
> > I added [Consensys/gnark-crypto#674](https://github.com/Consensys/gnark-crypto/pull/674) - I'll go over G2 map to curve to see that it is also useful for #1040.
> > Otherwise the map to curve also looks good. @yelhousni - please re-request review when you have addressed all the comments and I'll re-review. I had to merge master because there were some changes due to gnark-crypto API changes.
> 
> Addressed all the points except the `AssertIsOnCurve` and `AssertIsOnG1` in the `std/algebra.Curve` interface. No strong opinion here.

Yup - I think we could try to keep it simple for now. The only thing I'm worried about when having duplicate implementations (in sw_emulated and sw_bls12381) is that it may cause confusion when auditing or debugging bugs. I'll have another look, maybe I can come up with some idea while @ThomasPiellard is addressing the missing test comment.

ThomasPiellard commented on std/algebra/emulated/sw_bls12381/map_to_g1.go:
```diff
@@ -0,0 +1,190 @@
+	y1 := hint[0] // 18. (is_gx1_square, y1) = sqrt_ratio(tv2, tv6)
+
+	// TODO constrain gx1NSquare and y1
+	// (gx1NSquare==1 AND (u/v) QNR ) OR (gx1NSquare==0 AND (u/v) QR )
```
the TODO is done 👍

ivokub reviewed (approved):
Good from my side now. But before merging we need to wait for https://github.com/Consensys/gnark-crypto/pull/674 to be merged to gnark-crypto master and then we need to update the dependency in this PR. Otherwise we may have non-linear dependencies and depend on go package cache in case the gnark-crypto PR is deleted.

I'll try moving `AssertIsOnCurve` and `AssertIsOnG1` to interfaces and de-duplicate in separate PR.
